File: src/gazap-site/src/main/java/gazap/site/validation/ReCaptchaValidator.java
package gazap.site.validation;

import gazap.site.services.RecaptchaValidator;
import org.springframework.beans.BeanWrapper;
import org.springframework.beans.PropertyAccessorFactory;
import org.springframework.beans.factory.annotation.Autowired;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class ReCaptchaValidator implements ConstraintValidator<ReCaptcha, Object> {
    private String challengeField;
    private String responseField;

    @Autowired
    private RecaptchaValidator captchaValidator;

    @Override
    public void initialize(ReCaptcha ann) {
        challengeField = ann.challengeField();
        responseField = ann.responseField();
    }

    @Override
    public boolean isValid(Object value, ConstraintValidatorContext context) {
        BeanWrapper beanWrapper = PropertyAccessorFactory.forBeanPropertyAccess(value);
        String challenge = (String) beanWrapper.getPropertyValue(challengeField);
        String response = (String) beanWrapper.getPropertyValue(responseField);
        return captchaValidator.validate(challenge, response);
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/services/RecaptchaValidator.kt
package waypalm.site.services

import org.springframework.beans.factory.annotation.Value
import javax.inject.Named
import net.tanesha.recaptcha.ReCaptchaFactory

public trait RecaptchaValidator {
    fun validate(challenge: String, response: String): Boolean
}


Named
public class RecaptchaValidatorImpl : RecaptchaValidator {
    Value("\${recaptcha.public}") var publicKey: String = ""
    Value("\${recaptcha.private}")var privateKey: String = ""
    Value("\${recaptcha.site}")var site: String = ""
    Value("\${recaptcha.enabled}") var enabled: Boolean = false

    public override fun validate(challenge: String, response: String): Boolean {
        if (!enabled) return true;
        if (challenge.isEmpty() || response.isEmpty()) {
            return false;
        }
        return ReCaptchaFactory.newReCaptcha(publicKey, privateKey, false)!!
                .checkAnswer(site, challenge, response)!!
                .isValid();
    }
}
